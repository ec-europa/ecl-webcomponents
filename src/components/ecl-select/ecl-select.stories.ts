const getArgs = () => {
  return {
    required: true,
    disabled: false,
    invalid: false,
    width: 'm',
    label: 'Select a country',
    selectId: 'select-id',
    helperText: `This is the input helper text.`,
    requiredText: '*',
    invalidText: 'This is the error message',
    invalidIcon: 'error',
    optionalText: '(optional)',
    multiple: false,
    multiplePlaceholder: 'Select an item',
    multipleSearchNoResultsText: 'No results found',
    multipleAllText: 'Select all',
    multipleClearAllText: 'Clear all',
    multipleCloseText: 'Close',
    multipleSearchText: 'Enter keyword',
  };
};

const getArgTypes = () => {
  return {
    required: {
      type: { name: 'boolean' },
      description: 'Required input',
      table: {
        category: 'States',
      },
    },
    disabled: {
      type: { name: 'boolean' },
      description: 'Disabled input',
      table: {
        category: 'States',
      },
    },
    invalid: {
      type: { name: 'boolean' },
      description: 'Invalid input',
      table: {
        category: 'States',
      },
    },
    label: {
      type: { name: 'string' },
      description: 'Legend of the form group',
      table: {
        category: 'Content',
      },
    },
    requiredText: {
      name: 'required-text',
      type: { name: 'string' },
      description: 'Text to show when required',
      table: {
        category: 'Content',
      },
    },
    optionalText: {
      name: 'optional-text',
      type: { name: 'string' },
      description: 'Text to show when optional',
      table: {
        category: 'Content',
      },
    },
    helperText: {
      name: 'helper-text',
      type: { name: 'string' },
      description: 'Helper text',
      table: {
        category: 'Content',
      },
    },
    invalidText: {
      name: 'invalid-text',
      type: { name: 'string' },
      description: 'Error message',
      table: {
        category: 'Content',
      },
    },
    invalidIcon: {
      name: 'invalid-icon',
      type: { name: 'string' },
      description: 'Error icon',
      table: {
        category: 'Content',
      },
    },
    width: {
      type: { name: 'select' },
      description: 'Size of the select element',
      options: ['s', 'm', 'l'],
      table: {
        category: 'Styles',
      },
    },
    multiple: {
      name: 'multiple',
      type: { name: 'boolean' },
      description: 'Multiple select element',
      table: {
        category: 'multiple select',
      },
    },
    multiplePlaceholder: {
      name: 'multiple-placeholder',
      table: {
        category: 'multiple select',
      },
    },
    multipleSearchNoResultsText: {
      name: 'multiple-search-no-results-text',
      table: {
        category: 'multiple select',
      },
    },
    multipleAllText: {
      name: 'multiple-all-text',
      table: {
        category: 'multiple select',
      },
    },
    multipleClearAllText: {
      name: 'multiple-clear-all-text',
      table: {
        category: 'multiple select',
      },
    },
    multipleCloseText: {
      name: 'multiple-close-text',
      table: {
        category: 'multiple select',
      },
    },
    multipleSearchText: {
      name: 'multiple-search-text',
      table: {
        category: 'multiple select',
      },
    },
  };
};

export default {
  title: 'Components/form/select',
};

const Template = (args) => 
  `<ecl-form-group
    helper-text="${args.helperText}"
    helper-id="helper-id"
    theme="${args.theme}"
    invalid=${args.invalid}
    invalid-text="${args.invalidText}"
    invalid-icon="${args.invalidIcon}"
    optional-text="${args.optionalText}"
    label="${args.label}"
    required=${args.required}
    disabled="${args.disabled}"
  >
    <ecl-select
      required=${args.required}
      invalid=${args.invalid}
      disabled=${args.disabled}
      multiple=${args.multiple}
      multiple-placeholder="${args.multiplePlaceholder}"
      multiple-search-text="${args.multipleSearchText}"
      multiple-close-text="${args.multipleCloseText}"
      multiple-search-no-results-text="${args.multipleSearchNoResultsText}"
      multiple-all-text="${args.multipleAllText}"
      multiple-clear-all-text="${args.multipleClearAllText}"
      ecl-script=${args.multiple}
      name="ecl-select"
      theme="${args.theme}"
      width=${args.width}
    > 
      <optgroup label="European countries">
        <option value=1>Belgium</option>
        <option value=2>France</option>
        <option value=3 disabled>Luxembourg</option>
        <option value=4>Germany</option>
        <option value=5 selected>Bulgaria</option>
        <option value=6>Italy</option>
        <option value=7>Romania</option>
        <option value=8>Greece</option>
        <option value=9>Hungary</option>
        <option value=10>Portugal</option>
      </optgroup>
      <optgroup label="Non European countries">
        <option value=11>China</option>
        <option value=11>Turkey</option>
      </optgroup>
    </ecl-select>
  </ecl-form-group>`;

export const Select = Template.bind({});
Select.parameters = {
  viewMode: 'story',
};
Select.storyName = 'default';
Select.args = getArgs();
Select.argTypes = getArgTypes();