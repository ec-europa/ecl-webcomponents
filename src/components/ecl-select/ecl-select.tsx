import { Component, h, Prop, Element, getAssetPath, State, Event, EventEmitter } from '@stencil/core';
declare const ECL: any;

@Component({
  tag: 'ecl-select',
  styleUrls: {
    ec: './build/styles/ecl-select-ec.css',
    eu: './build/styles/ecl-select-eu.css',
  },
  shadow: false,
  scoped: false,
  assetsDirs: ['build'],
})

export class EclSelect {
  @Element() el: HTMLElement;
  @Prop() theme: string = 'ec';
  @Prop() styleClass: string;
  @Prop() eclScript: boolean = false;
  @Prop() disabled: boolean = false;
  @Prop() required: boolean = false;
  @Prop() invalid: boolean = false;
  @Prop() width: string = 'm';
  @Prop() selectId: string;
  @Prop() name: string;
  @Prop() multiple: boolean = false;
  @Prop() multiplePlaceholder: string;
  @Prop() multipleSearchText: string;
  @Prop() multipleSearchNoResultsText: string;
  @Prop() multipleAllText: string;
  @Prop() multipleClearAllText: string;
  @Prop() multipleCloseText: string;
  @State() value: string;
  @Prop() hasChanged: boolean = false;
  @Prop() isFocused: boolean = false;
  @Event() inputFocus: EventEmitter<FocusEvent>;
  @Event() inputBlur: EventEmitter<FocusEvent>;
  @Event() inputChange: EventEmitter;

  getClass(): string {
    const styleClasses = [
      `ecl-select__container`,
      `ecl-select__container--${this.width}`,
      this.styleClass
    ];

    if (this.disabled) {
      styleClasses.push('ecl-select__container--disabled');
    }
    if (this.invalid) {
      styleClasses.push('ecl-select__container--invalid');
    }

    return styleClasses.join(' ');
  }

  componentDidLoad() {
    if (this.eclScript && this.multiple) {
      // Load the ECL vanilla js if not already present.
      const src = getAssetPath('./build/scripts/ecl-select-vanilla.js');
      if (document.querySelector(`script[src="${src}"]`)) {
        document.querySelector(`script[src="${src}"]`).remove();
      }
      const script = document.createElement('script');
      script.src = src;
      script.onload = () => {
        const select = new ECL.Select(this.el.getElementsByTagName('select')[0]);
        select.init();
      };

      document.body.appendChild(script);
    }
  }

  handleInput(event) {
    this.value = event.target.value;
  }

  handleFocus(event) {
    this.inputFocus.emit(event);
    this.isFocused = true;
  }

  handleChange(event) {
    this.inputChange.emit(event);
    this.hasChanged = true;
  }

  handleBlur(event) {
    this.inputBlur.emit(event);
    this.isFocused = false;
  }

  render() {
    let attributes = {
      class: 'ecl-select',
      id: this.selectId,
      name: this.name,
      required: this.required,
      disabled: this.disabled,
    };

    if (this.multiple) {
      attributes['multiple'] = 'data-ecl-select-multiple';
      attributes['data-ecl-select-default'] = this.multiplePlaceholder;
      attributes['data-ecl-select-no-results'] = this.multipleSearchNoResultsText;
      attributes['data-ecl-select-all'] = this.multipleAllText;
      attributes['data-ecl-select-clear-all'] = this.multipleClearAllText;
      attributes['data-ecl-select-close'] = this.multipleCloseText;
      attributes['data-ecl-select-search'] = this.multipleSearchText;
    }

    return (
      <div 
        class={this.getClass()}
      >
        <select
          {...attributes}
          onInput={ev => this.handleInput(ev)}
          onFocus={ev => this.handleFocus(ev)}
          onBlur={ev => this.handleBlur(ev)}
          onChange={ev => this.handleChange(ev)}
        >
          <slot></slot>
        </select>
        <div class="ecl-select__icon">
          <ecl-icon
            style-class={`ecl-select__icon-shape sc-ecl-select-${this.theme}`}
            icon="corner-arrow"
            size="s"
            transform="rotate-180"
            ></ecl-icon>
        </div>
      </div>
    );
  }
}