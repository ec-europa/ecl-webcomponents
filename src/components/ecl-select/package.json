{
  "name": "@ecl/web-components-select",
  "author": "European Commission",
  "license": "EUPL-1.2",
  "version": "0.0.0",
  "description": "ECL Select",
  "main": "ecl-select.tsx",
  "private": true,
  "scripts": {
    "build:scripts": "cross-env ecl-builder scripts",
    "build:styles": "cross-env ecl-builder styles",
    "build": "run-s clean build:*",
    "clean": "rimraf build",
    "prepublish": "npm run build"
  },
  "devDependencies": {
    "@ecl/builder": "3.6.0",
    "@ecl/vanilla-component-select": "3.6.0"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/planctus/ecl-webcomponents.git"
  },
  "bugs": {
    "url": "https://github.com/planctus/ecl-webcomponents/issues"
  },
  "homepage": "https://github.com/planctus/ecl-webcomponents",
  "keywords": [
    "ecl",
    "europa-component-library",
    "design-system",
    "web-components"
  ]
}