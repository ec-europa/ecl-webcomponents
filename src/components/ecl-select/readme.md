# ecl-select



<!-- Auto Generated Below -->


## Properties

| Property                      | Attribute                         | Description | Type      | Default     |
| ----------------------------- | --------------------------------- | ----------- | --------- | ----------- |
| `disabled`                    | `disabled`                        |             | `boolean` | `false`     |
| `eclScript`                   | `ecl-script`                      |             | `boolean` | `false`     |
| `hasChanged`                  | `has-changed`                     |             | `boolean` | `false`     |
| `inputValue`                  | `input-value`                     |             | `string`  | `undefined` |
| `invalid`                     | `invalid`                         |             | `boolean` | `false`     |
| `multiple`                    | `multiple`                        |             | `boolean` | `false`     |
| `multipleAllText`             | `multiple-all-text`               |             | `string`  | `undefined` |
| `multipleClearAllText`        | `multiple-clear-all-text`         |             | `string`  | `undefined` |
| `multipleCloseText`           | `multiple-close-text`             |             | `string`  | `undefined` |
| `multiplePlaceholder`         | `multiple-placeholder`            |             | `string`  | `undefined` |
| `multipleSearchNoResultsText` | `multiple-search-no-results-text` |             | `string`  | `undefined` |
| `multipleSearchText`          | `multiple-search-text`            |             | `string`  | `undefined` |
| `name`                        | `name`                            |             | `string`  | `undefined` |
| `required`                    | `required`                        |             | `boolean` | `false`     |
| `selectId`                    | `select-id`                       |             | `string`  | `undefined` |
| `styleClass`                  | `style-class`                     |             | `string`  | `undefined` |
| `theme`                       | `theme`                           |             | `string`  | `'ec'`      |
| `width`                       | `width`                           |             | `string`  | `'m'`       |


## Events

| Event         | Description | Type                      |
| ------------- | ----------- | ------------------------- |
| `inputBlur`   |             | `CustomEvent<FocusEvent>` |
| `inputChange` |             | `CustomEvent<any>`        |
| `inputFocus`  |             | `CustomEvent<FocusEvent>` |


## Dependencies

### Depends on

- [ecl-icon](../ecl-icon)

### Graph
```mermaid
graph TD;
  ecl-select --> ecl-icon
  style ecl-select fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
