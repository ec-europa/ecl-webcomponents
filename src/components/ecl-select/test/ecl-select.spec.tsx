import { newSpecPage } from '@stencil/core/testing';
import { EclSelect } from '../ecl-select';

describe('ecl-select', () => {
  it('renders', async () => {
    const page = await newSpecPage({
      components: [EclSelect],
      html: `<ecl-select></ecl-select>`,
    });
    expect(page.root).toEqualHtml(`
      <ecl-select>
        <mock:shadow-root>
          <slot></slot>
        </mock:shadow-root>
      </ecl-select>
    `);
  });
});
