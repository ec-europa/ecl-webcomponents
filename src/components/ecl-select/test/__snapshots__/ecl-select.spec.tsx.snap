// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ecl-select renders 1`] = `
<ecl-form-group helper-id="helper-id" helper-text="Helper text" label="Select input field" required="">
  <!---->
  <fieldset aria-describedby="helper-id" class="ecl-form-group">
    <legend class="ecl-form-label">
      Select input field
      <span class="ecl-form-label__required">
        *
      </span>
    </legend>
    <div class="ecl-help-block">
      Helper text
    </div>
    <ecl-select name="ecl-select" required="" width="s">
      <!---->
      <div class="ecl-select__container ecl-select__container--s">
        <select class="ecl-select" name="ecl-select" required="">
          <option value="1">
            Belgium
          </option>
          <option value="2">
            France
          </option>
          <option disabled="" value="3">
            Luxembourg
          </option>
          <option value="4">
            Germany
          </option>
          <option selected="" value="5">
            Bulgaria
          </option>
          <option value="6">
            Italy
          </option>
          <option value="7">
            Romania
          </option>
          <option value="8">
            Greece
          </option>
          <option value="9">
            Hungary
          </option>
          <option value="10">
            Portugal
          </option>
        </select>
        <div class="ecl-select__icon">
          <ecl-icon icon="corner-arrow" size="s" style-class="ecl-select__icon-shape sc-ecl-select-ec" transform="rotate-180"></ecl-icon>
        </div>
      </div>
    </ecl-select>
  </fieldset>
</ecl-form-group>
`;
